Promise/Promise+Extras.swift

```diff
@@ -150,10 +150,19 @@ extension Promise {
     }
 
     public func recover(_ recovery: @escaping (Error) throws -> Promise<Value>) -> Promise<Value> {
+        return recover(type: Error.self, recovery)
+    }
+
+    public func recover<E: Error>(type errorType: E.Type, _ recovery: @escaping (E) throws -> Promise<Value>) -> Promise<Value> {
         return Promise(work: { fulfill, reject in
             self.then(fulfill).catch({ error in
+                guard let castError = error as? E else {
+                    reject(error)
+                    return
+                }
+
                 do {
-                    try recovery(error).then(fulfill, reject)
+                    try recovery(castError).then(fulfill, reject)
                 } catch (let error) {
                     reject(error)
```

Comment:
This line is not explicitly tested with the new tests, because it's already covered by existing ones.